spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.reactive.LoadBalancerProperties;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that verifies whether the
+ * instances are alive and only returns the healthy one, unless there are none. Uses
+ * {@link WebClient} to ping the <code>health</code> endpoint of the instances.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 2.2.0
+ */
+public class HealthCheckServiceInstanceListSupplier
+		implements ServiceInstanceListSupplier {
+
+	private static final Log LOG = LogFactory
+			.getLog(HealthCheckServiceInstanceListSupplier.class);
+
+	private final ServiceInstanceListSupplier delegate;
+
+	private final LoadBalancerProperties.HealthCheck healthCheck;
+
+	private final WebClient webClient;
+
+	private final String defaultHealthCheckPath;
+
+	private List<ServiceInstance> instances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	private List<ServiceInstance> healthyInstances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	public HealthCheckServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			LoadBalancerProperties.HealthCheck healthCheck, WebClient webClient) {
+		this.delegate = delegate;
+		this.healthCheck = healthCheck;
+		defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
+				"/actuator/health");
+		this.webClient = webClient;
+		initInstances();
+
+	}
+
+	private void initInstances() {
+		delegate.get().subscribe(delegateInstances -> {
+			instances.clear();
+			instances.addAll(delegateInstances);
+		});
+
+		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
+
+		healthCheckFlux.subscribe(verifiedInstances -> {
+			healthyInstances.clear();
+			healthyInstances.addAll(verifiedInstances);
+		});
+	}
+
+	protected Flux<List<ServiceInstance>> healthCheckFlux() {
+		return Flux.create(emitter -> Schedulers
+				.newSingle("Health Check Verifier: " + getServiceId(), true)
+				.schedulePeriodically(() -> {
```

Comment:
Task is never disposed  and cannot be disposed.  
What if one wants to refresh/recreate `HealthCheckServiceInstanceListSupplier`  bean. It could lead to leak and unnecessary pinging.